src/corvus.c

```diff
@@ -134,30 +138,58 @@ int config_add(char *name, char *value)
             memcpy(config.requirepass, value, strlen(value));
         }
     } else if (strcmp(name, "node") == 0) {
-        cv_free(config.node.addr);
-        memset(&config.node, 0, sizeof(config.node));
-
+    	struct address *addr = NULL;
+    	int addr_cnt = 0;
         char *p = strtok(value, ",");
         while (p) {
-            config.node.addr = cv_realloc(config.node.addr,
-                    sizeof(struct address) * (config.node.len + 1));
-
-            if (socket_parse_ip(p, &config.node.addr[config.node.len]) == -1) {
-                cv_free(config.node.addr);
-                return -1;
+        	addr = cv_realloc(addr, sizeof(struct address) * (addr_cnt + 1));
+            if (socket_parse_ip(p, &addr[addr_cnt]) == -1) {
+                cv_free(addr);
+                return CORVUS_ERR;
             }
-
-            config.node.len++;
+            addr_cnt++;
             p = strtok(NULL, ",");
         }
+		{
+			struct node_conf *newnode = cv_calloc(1, sizeof(struct node_conf));
+			memset(newnode, 0, sizeof(struct node_conf));
+			newnode->addr = addr;
+			newnode->len = addr_cnt;
+			newnode->refcount = 1;
+			struct node_conf *oldnode = config.node;
+			config.node = newnode;
+			conf_node_dec_ref(oldnode);
+		}
     } else if (strcmp(name, "slowlog-log-slower-than") == 0) {
         config.slowlog_log_slower_than = atoi(value);
     } else if (strcmp(name, "slowlog-max-len") == 0) {
         config.slowlog_max_len = atoi(value);
     } else if (strcmp(name, "slowlog-statsd-enabled") == 0) {
         config.slowlog_statsd_enabled = atoi(value);
     }
-    return 0;
+    return CORVUS_OK;
+}
+
+struct node_conf *conf_node_inc_ref()
+{
+    pthread_mutex_lock(&lock_conf_node);
+    struct node_conf *node = config.node;
+    int refcount = ATOMIC_INC(node->refcount, 1);
+    assert(refcount >= 0);
+    pthread_mutex_unlock(&lock_conf_node);
+    return node;
+}
+
+void conf_node_dec_ref(struct node_conf *node)
+{
+    pthread_mutex_lock(&lock_conf_node);
+    int refcount = ATOMIC_DEC(node->refcount, 1);
```

Comment:
I think we can safely remove the lock here. Thanks to the atomic operation multiple calling of `conf_node_dec_ref` will end up with only one zero `refcount`;